### PR TITLE
Fast 2d convex hull

### PR DESCRIPTION
**Pull request by lmmx**
## Description

Bug fix for the patch submitted by @ehusby in #2928 3 years ago (which this PR supersedes), following advice from @jni to send in a fresh PR (as I don't have merge permissions so can't append to that PR as I previously tried to)

- The original PR computed a faster convex hull (see proposal in that thread for details).
- There was discussion about controlling whether to use the fast version with a flag, but this was not followed up on, which I now implement in this PR (I cannot find a way to include his commits unfortunately :disappointed:)

My contribution is to fix a bug in the application of offsets to the computed convex hull (specifically an `IndexError` which arose when a half-pixel was rounded to a whole pixel and this whole pixel was one pixel outside of the perimeter of the image, and thus outside of the mask made from the image, which it tried to index into resulting in an error).

- For a full description of the source of the bug see [my comment in the previous PR thread](https://github.com/scikit-image/scikit-image/pull/2928#issuecomment-735432524)

Further to this, I have added in conditional checks so that this new routine is only used in the 2D case, as the previous PR would have applied to 3D cases too from what I can see (and I am under the impression it is explicitly only intended for 2D hulls, as the title of the PR stated).

While developing the modification I inspected the source of the bug (during and in the months following the 2020 SciPy sprint), in [this repo](https://github.com/lmmx/fast-convex-hull-2d/), specifically [the script](https://github.com/lmmx/fast-convex-hull-2d/blob/master/partial_coord_offset.py) `partial_coord_offset.py` reproduced the bug and then remediated it.

Finally I ran the test suite in pytest, and it broke the majority of the convex hull-related tests and some others (detailed below).

Fixing the IndexError was the first step in @jni's to-do list from 2018, and upon reviewing this I have now also implemented the 2nd step (which involved re-inserting the code lines which Erik had originally removed, so that we get continuous performance and the pre-existing tests will still pass for now):

> - [x] fix the errors.
> - [x] add a keyword argument to `convex_hull_image`. Naming suggestion: `fast_drawing=None`, which gets converted to False by default in 0.14 (**edit**: this would now be 0.19), and True from 0.16 onwards (**edit**: this would now be 0.21). See the deprecation cycle guidelines from the contributor guide. After 0.16 we can deprecate the argument altogether.
> - [ ] add tests that show the correct (pixel-wise) behaviour for `fast_drawing=True`, and pass `fast_drawing=False` for existing tests.

So this just leaves the tests to write for the new behaviour, and these can be handled to transition from the old routine to the new one by using the `fast_drawing` parameter.

Additionally, I will need to add the deprecation to the TODO list, I suggest:

```rst
Version 0.21
------------
In ``skimage/morphology/convex_hull.py``, set the `fast_drawing` default value to 'True' in ``convex_hull_image``
```

Additionally, it should be deprecated by removal in v0.22

```rst
Version 0.22
------------
In ``skimage/morphology/convex_hull.py``, remove the `fast_drawing` argument from ``convex_hull_image``
```

I haven't considered deprecation warnings yet, so will look into this (any guidance welcome but I expect it will be documented in `CONTRIBUTING.txt` when I look more closely). Can someone confirm I got that right (about the TODO items and which sections they go in, and the need for deprecation warnings) please?

## Checklist

- [x] Docstrings
  - I introduced a single 'internal' funcdef (i.e. whose name is prefixed by a single underscore) and gave a simple docstring explaining what it does but not the extent of a 'public-facing' funcdef in the API like `convex_hull_image` has.
- [x] Gallery example in `./doc/examples` (already exists, as the feature is not new, just speeds up an existing feature)
  - `doc/examples/edges/plot_convex_hull.py`
  - Running the example (before adding the deprecation, i.e. with the new routine), the output's result looks perfect (the difference shows the convex hull of the horse fits tightly)
- [ ] Benchmark in `./benchmarks`
  - existing feature so expect to be covered in the morphology benchmark
  - I'm not familiar with how these work, though I think I saw the web interface the other month (could someone point me to how I check for a regression/speedup?)
- [ ] Unit test
  - existing feature so already unit tested, but 6 of 8 are failing (and must be rewritten I presume)
  - `skimage/morphology/tests/test_convex_hull.py`
    - after adding old routine to deprecate: `skimage/morphology/tests/test_convex_hull.py ........            [ 83%]`
    - before adding the old routine back in: `skimage/morphology/tests/test_convex_hull.py F.FF.FFF            [ 83%]`

Without deprecation:
  
```
=================short test summary info ===================
FAILED skimage/feature/tests/test_censure.py::test_keypoints_censure_moon_image_octagon - AssertionError: 
FAILED skimage/feature/tests/test_censure.py::test_keypoints_censure_moon_image_star - AssertionError: 
FAILED skimage/measure/tests/test_regionprops.py::test_all_props_3d - ValueError: too many values to unpack (expected 2)
FAILED skimage/measure/tests/test_regionprops.py::test_feret_diameter_max_3d - ValueError: too many values to unpack (expected 2)
FAILED skimage/measure/tests/test_regionprops.py::test_convex_area - assert 133 == 125
FAILED skimage/measure/tests/test_regionprops.py::test_convex_image - AssertionError: 
FAILED skimage/measure/tests/test_regionprops.py::test_solidity - AssertionError: 
FAILED skimage/morphology/tests/test_convex_hull.py::test_basic - AssertionError: 
FAILED skimage/morphology/tests/test_convex_hull.py::test_qhull_offset_example - AssertionError: 
FAILED skimage/morphology/tests/test_convex_hull.py::test_pathological_qhull_example - AssertionError: 
FAILED skimage/morphology/tests/test_convex_hull.py::test_object - AssertionError: 
FAILED skimage/morphology/tests/test_convex_hull.py::test_non_c_contiguous - ValueError: too many values to unpack (expected 2)
FAILED skimage/morphology/tests/test_convex_hull.py::test_consistent_2d_3d_hulls - ValueError: too many values to unpack (expected 2)
FAILED skimage/morphology/tests/test_grey.py::TestMorphology::test_gray_morphology - AssertionError: 
FAILED skimage/morphology/tests/test_selem.py::TestSElem::test_selem_octagon - AssertionError: 
===================================================================================================== 15 failed, 6297 passed, 31 skipped, 14 warnings
```

With deprecation:

```
=========== warnings summary ===========
skimage/filters/rank/generic.py:1
  /home/louis/dev/skimage-pr-new/skimage/filters/rank/generic.py:1: DeprecationWarning: invalid escape sequence \-
    """

../../miniconda/envs/skimage-final-pr/lib/python3.8/importlib/__init__.py:127
  /home/louis/miniconda/envs/skimage-final-pr/lib/python3.8/importlib/__init__.py:127: UserWarning: Viewer requires Qt
    return _bootstrap._gcd_import(name[level:], package, level)

skimage/io/tests/test_tifffile.py::test_tifffile_kwarg_passthrough
  /home/louis/miniconda/envs/skimage-final-pr/lib/python3.8/site-packages/tifffile/tifffile.py:2835: DeprecationWarning: TiffFile: the 'fastij' argument is ignored
    warnings.warn(

skimage/io/tests/test_tifffile.py::test_tifffile_kwarg_passthrough
  /home/louis/miniconda/envs/skimage-final-pr/lib/python3.8/site-packages/tifffile/tifffile.py:2835: DeprecationWarning: TiffFile: the 'multifile' argument is ignored
    warnings.warn(

skimage/io/tests/test_tifffile.py::test_tifffile_kwarg_passthrough
  /home/louis/miniconda/envs/skimage-final-pr/lib/python3.8/site-packages/tifffile/tifffile.py:2835: DeprecationWarning: TiffFile: the 'multifile_close' argument is ignored
    warnings.warn(

skimage/measure/tests/test_fit.py::test_ransac_is_data_valid
skimage/measure/tests/test_fit.py::test_ransac_is_model_valid
skimage/measure/tests/test_fit.py::test_ransac_sample_duplicates
  /home/louis/dev/skimage-pr-new/skimage/measure/fit.py:879: UserWarning: No inliers found. Model not fitted
    warn("No inliers found. Model not fitted")

skimage/restoration/tests/test_restoration.py::test_richardson_lucy_filtered[float32-float32]
skimage/restoration/tests/test_restoration.py::test_richardson_lucy_filtered[float32-float64]
skimage/restoration/tests/test_restoration.py::test_richardson_lucy_filtered[float64-float32]
skimage/restoration/tests/test_restoration.py::test_richardson_lucy_filtered[float64-float64]
  /home/louis/dev/skimage-pr-new/skimage/restoration/deconvolution.py:376: RuntimeWarning: invalid value encountered in true_divide
    relative_blur = np.where(conv < filter_epsilon, 0, image / conv)

skimage/util/tests/test_arraypad.py::TypeError1::test_check_wrong_pad_amount
  /home/louis/miniconda/envs/skimage-final-pr/lib/python3.8/site-packages/numpy/core/_asarray.py:83: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
    return array(a, dtype, copy=False, order=order)

-- Docs: https://docs.pytest.org/en/stable/warnings.html
================= 6312 passed, 31 skipped, 13 warnings
```

- [ ] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)

In conclusion, I need to next work on the tests to advance this PR to a stage at which it can be reviewed but want to submit at this point to initiate the process.

## For reviewers

- Check that the PR title is short, concise, and will make sense 1 year
  later.
- Check that new functions are imported in corresponding `__init__.py`.
- Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.

### CODE REVIEW

File: skimage/morphology/convex_hull.py [hunk 1]
```diff
@@ -88,6 +107,11 @@ def convex_hull_image(image, offset_coordinates=True, tolerance=1e-10):
         Tolerance when determining whether a point is inside the hull. Due
         to numerical floating point errors, a tolerance of 0 can result in
         some points erroneously being classified as being outside the hull.
+    fast_drawing : bool or None, optional
+        If ``True``, use the fast SciPy function ``polygon_perimeter`` to locate
```
Comment (emmanuelle) [inline skimage/morphology/convex_hull.py:111]: not sure I understand here, `polygon_perimeter` is a scikit-image function, not a scipy one? I had the impression that the fast scipy function was `binary_fill_holes`.

File: skimage/morphology/convex_hull.py [hunk 2]
```diff
@@ -17,6 +19,23 @@ def _offsets_diamond(ndim):
         offsets[vertex, axis] = offset
     return offsets
 
+def _apply_partial_offsets(img, coords, offsets):
+    """
+    Apply the offsets only to the non-edge pixels, along with the trivial zero-offset.
+    """
```
Comment (emmanuelle) [inline skimage/morphology/convex_hull.py:25]: This function is a bit hard to read, could you please add a minimal example here in the docstring?

**Comment by emmanuelle** [comment]
Hi @lmmx thanks a lot for the PR! A couple of points:
- for benchmarks please take a look at https://scikit-image.org/docs/dev/contribute.html#benchmarks
- for the technical aspects of deprecation, you're indeed on the right track (items to add to the TODO file, plus warnings to raise in the function. You can also use of the decorators defined in `skimage._shared.utils` such as `deprecate_kwarg`). 

However, we first need to decide what / if we should deprecate. In #2928 there were a couple of tests on simple and more complex images which showed that there were some differences between the old and the new mode, both for `offset_coordinates=True` and `False`. From a test on the simple image
```python
import numpy as np
bw = np.array([
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 1, 0, 0, 0, 0],
    [0, 0, 0, 1, 0, 1, 0, 0, 0],
    [0, 0, 1, 0, 0, 0, 1, 0, 0],
    [0, 1, 0, 0, 0, 0, 0, 1, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    ])
out_1 = morphology.convex_hull_image(bw).astype(np.uint8)
out_2 = morphology.convex_hull_image(bw, fast_drawing=True, offset_coordinates=False).astype(np.uint8)
```
I got the impression that your bug fix had fixed the difference between the two modes for `offset_coordinates=False`. Therefore I wonder if what we should deprecate is `offset_coordinates` maybe? In the test suite there is no example/test showing how the value of this parameter would change the output with the old mode. Anyway, some simple examples showing the different cases would help for the review.

File: skimage/morphology/convex_hull.py [hunk 3]
```diff
@@ -17,6 +19,23 @@ def _offsets_diamond(ndim):
...
+    # Insert the trivial offset of `[0., 0.]` into `offsets` i.e. retain original points
+    offsets = np.insert(offsets, 0, 0., axis=0)
+    row_max, col_max = np.subtract(img.shape, 1)
+    # bool masks for the subsets of `coords` including each edge (one edge at a time)
+    edge_t, edge_b = [coords[:,0] == lim for lim in (0, row_max)]
+    edge_l, edge_r = [coords[:,1] == lim for lim in (0, col_max)]
+    edge_includers = [edge_t, edge_b, edge_l, edge_r]
+    dummy_edge = np.zeros_like(edge_t, dtype=bool) # all-False so offset always applied
+    edge_includers.insert(0, dummy_edge)
```
Comment (rfezzani) [inline skimage/morphology/convex_hull.py:34]:
```suggestion
    edge_includers = np.insert(np.equal(np.repeat(coords, 2, axis=1),
                                        [0, row_max, 0, col_max]),
                                0, False, axis=1)
```

File: skimage/morphology/convex_hull.py [hunk 4]
```diff
@@ -88,6 +107,11 @@ def convex_hull_image(image, offset_coordinates=True, tolerance=1e-10):
         Tolerance when determining whether a point is inside the hull. Due
         to numerical floating point errors, a tolerance of 0 can result in
         some points erroneously being classified as being outside the hull.
+    fast_drawing : bool or None, optional
+        If ``True``, use the fast SciPy function ``polygon_perimeter`` to locate
+        hull perimeter pixels and then fill it in. If ``False``, use the slower
+        Cython function ``grid_points_in_poly`` to locate convex hull pixels.
+        If ``None``, default to ``False`` (in future will default to ``True``).
```
Comment (rfezzani) [inline skimage/morphology/convex_hull.py:114]: We need to specify the version when the change will be effective.

**Review by rfezzani — COMMENTED** [review]
Thank you @lmmx! Can you please add a benchmark asserting the acceleration that you claim?

**Comment by ehusby** [comment]
After starting the work this PR is based on back in 2017, I'm now circling back to more work in the large image manipulation field. Would love to see this PR followed up on! @lmmx I'd back happy to take back the reins on this one if the patches we've assembled here are still applicable.

**Comment by lmmx** [comment]
oh wow pandemic throwback, yes I am totally down for that let’s go! 😀 at a glance the bench and unit tests are left to do, and some new conflicts to resolve (as you might well expect!)

**Comment by lagru** [comment]
Yes, please go for it! Feel welcome to continue here or in a new PR

Note that we haven't been doing a good job of maintaining our benchmark infrastructure. :see_no_evil: And I don't think we've been running them consistently. So we may need to verify performance gains on our local machines for now. But it would still be good to have a benchmark implemented for this in case we get around to running them regularly again.

**Comment by matthew-brett** [comment]
We have changed the base branch for this PR from the `main` branch to the `pre-megamove-integration` branch.

## Summary - what this means for your PR

You can continue to work on your PR, now based on `pre-megamove-integration`.  You should update your branch to the current state of `pre-megamove-integration` (this is what the Github "Update branch" button does).   Please **do not merge the scikit-image `main` branch into your PR branch**.

Once we've merged your PR into the `pre-megamove-integration` branch, we'll update your changes for Megamove, and merge into the `main` branch.

In addition - our automated script detected that your branch had merge conflicts with the pre-Megamove branch. Please make sure to merge in the `pre-megamove-integration` branch, and resolve the merge conflicts in the way that you want.

Read on for the Megamove rationale, and more detailed instructions.

<details>

## Rationale

We've changed the base branch to make it easier and smoother to continue to work on your PR, while we work through a rather large change in the directory structure of the repository - a change that we have called *Megamove*.

See [the Megamove explanation issue](https://github.com/scikit-image/scikit-image/issues/8232) for more detail.

Changing the base branch means you do not see merge conflicts related to the *Megamove* change, and can continue to work with the repository in its original state, without worrying about porting those changes across to the new post-Megamove structure.

As you'll see in the linked issue, you can continue to work on your PR as before, and you can continue to use the Github GUI to "Update branch" - this updates from the new `pre-megamove-integration` branch.   However, there is one caveat: please, do not try and merge the `scikit-image` `main` branch into this one, but instead, keep track of the `pre-megamove-integration` branch.  The issue explains more.

To emphasize again, **do not try and merge the scikit-image `main` branch**.  You will get Megamove related merge conflicts that are rather tricky to resolve correctly.

When we've approved your PR, we'll merge into the `pre-megamove-integration` branch, and then process it so we can merge it into the `main` branch.   We'll let you know when we've merged your PR changes into `main`.

</details>

Any questions, please reply here.
